kernel-modules/zephyr/zephyr-directory-scanner/src/main/java/io/zephyr/scan/DirectoryScanner.java

```diff
@@ -0,0 +1,240 @@
+      registerKeys(zephyr, watchService);
+      watch(zephyr, watchService);
+    }
+  }
+
+  private void watch(Zephyr zephyr, WatchService watchService) {
+    while (running) {
+      try {
+        val key = watchService.take();
+        doHandle(zephyr, key, watchService);
+      } catch (InterruptedException ex) {
+        ex.printStackTrace();
+        return;
+      }
+    }
+  }
+
+  private void doHandle(Zephyr zephyr, WatchKey key, WatchService watchService) {
+    for (val event : key.pollEvents()) {
+      handleEvent(zephyr, key, event, watchService);
+    }
+  }
+
+  private void registerKeys(Zephyr zephyr, WatchService watchService) throws IOException {
+    val paths = options.getDirectories();
+    if (paths == null || paths.length == 0) {
+      register(zephyr, fileSystem.getPath("deployments"), watchService);
+      return;
+    } else {
+      for (val path : paths) {
+        register(zephyr, Path.of(path), watchService);
+      }
+    }
+  }
+
+  private void register(Zephyr zephyr, Path path, WatchService watchService) throws IOException {
+    log.log(Level.INFO, "scanner.watching.path", path);
+    val key = path.register(watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+    keys.put(key, path);
+
+    if (options.isInstallOnStart()) {
+      val file = path.toFile();
+      val files = file.listFiles();
+      if (files != null) {
+        val installables =
+            Arrays.stream(files)
+                .map(
+                    t -> {
+                      try {
+                        return t.getAbsoluteFile().toURI().toURL();
+                      } catch (MalformedURLException ex) {
+                        log.log(Level.WARNING, "scanner.directory.scan.failed", ex);
+                        return null;
+                      }
+                    })
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        zephyr.install(installables);
+      }
+    }
+  }
+
+  public void stop() {
+
+    try {
+      if (watchService != null) {
+        watchService.close();
+      }
+    } catch (IOException e) {
+      log.log(Level.WARNING, "scanner.watchservice.close.failed", e);
+    } finally {
+      running = false;
+    }
+  }
+
+  private boolean exportsKernel(EntryPoint entryPoint) {
+    return entryPoint.getService(Kernel.class) != null;
+  }
+
+  @Override
+  public Logger getLogger() {
+    return log;
+  }
+
+  @Override
+  public DirectoryScannerOptions getOptions() {
+    return options;
+  }
+
+  @Override
+  public int getPriority() {
+    return 100;
+  }
+
+  private void handleEvent(
+      Zephyr zephyr, WatchKey key, WatchEvent<?> event, WatchService watchService) {
+    if (event.kind() == ENTRY_CREATE) {
+      handleCreate(zephyr, key, event, watchService);
+    }
+    if (event.kind() == ENTRY_DELETE) {
+      handleDelete(zephyr, key, event, watchService);
+    }
+    if (event.kind() == ENTRY_MODIFY) {
```

Comment:
shouldn't this be handleModify()? or at the very least handleDelete() and then handleCreate()?

Reply:
Linux seems to be firing MODIFY instead of CREATE when a file is created.